lib/lrmd/lrmd_alerts.c

```diff
@@ -176,6 +179,12 @@ exec_alert_list(lrmd_t *lrmd, GList *alert_list, enum crm_alert_flags kind,
                                               timestamp);
                 free(timestamp);
             }
+
+            crm_time_set_hr_dt(dt, now);
+            snprintf(timestamp_epoch, 32, "%lld", crm_time_get_seconds_since_epoch(dt));
```

Comment:
I'd rather avoid this function, because I don't think it's 100% accurate. See crm_now_string() in tools/crm_mon.c for something similar that uses only time() and ctime(). We would just have to strdup() the result.

Reply:
I will check crm_now_string ().

Reply:
> Does your opinion mean the following things?
> 
>     change crm_time_get_seconds_since_epoch() -> time().
>     ctime() does not use it because it generates a format with a script(alert_snmp.sh.sample).
> 
> 
> snprintf(timestamp_epoch, 32, "%lld", time());
> 
> If so, there will be a difference between the time earlier acquired with crm_time_hr_new () and the time taken with this time ().
> 
> static int
> exec_alert_list(lrmd_t *lrmd, GList *alert_list, enum crm_alert_flags kind,
>                 const char *attr_name, lrmd_key_value_t *params)
> (snip)
>         if (now == NULL) {
>             now = crm_time_hr_new(NULL);
>         }
>         crm_info("Sending %s alert via %s to %s",
>                  kind_s, entry->id, entry->recipient);
> (snip)
>                  snprintf(timestamp_epoch, 32, "%lld", time());
> 
> With the same crm file, the difference in time is not an issue when setting FILE or SMTP and SNMP Alert?
> 
> ex.
> (snip)
> alert alert_1 /usr/share/pacemaker/alerts/alert_snmp.sh.sample \
>      attributes \
>          trap_add_hires_timestamp_oid="true" \
>          trap_resource_tasks="start,stop,monitor,promote,demote" \
>      to 192.168.40.1
> ---->(The time of time() is sent to execute this Alert)
> alert alert_2 /usr/share/pacemaker/alerts/alert_file.sh.sample \
>      attributes \
>          trap_resource_tasks="start,stop,monitor,promote,demote" \
>      to /tmp/test.log
> ----->(The time of crm_time_hr_new() is sent to execute this Alert)
> (snip)
> 
> Even if you do not care about the time gap....
> Since it is necessary to pass milliseconds of data to the SNMP script, is not gettimeofday () rather than time ()?
> (Actually SNMP DateAndTime is up to "deci seconds".)
> 
> 
> gettimeofday(&snmp_now, NULL);
> snprintf(timestamp_epoch, 32, "%lld", snmp_now.tv_sec);
> snprintf(timestamp_usec, 12, "%06d", snmp_now.tv_usec);

Ah, yes, you are completely correct!

> Do not you think it would be better to modify crm_time_hr_new () and return the value obtained with gettimeofday () and process it?
> 
> ex:
> crm_time_hr_t *
> crm_time_hr_new(const char *date_time, struct timeval *get_tv_now)
> {
>     crm_time_hr_t *hr_dt = NULL;
>     struct timeval tv_now;
> 
>     if (!date_time) {
>         if (gettimeofday(&tv_now, NULL) == 0) {
>             if (get_tv_now != NULL) {
> 		*get_tv_now = tv_now;
> 	    }
>             hr_dt = crm_time_timeval_hr_convert(NULL, &tv_now);
>         }
> (snip)

That won't be necessary. We can call gettimeofday() as you suggested earlier, and then use crm_time_timeval_hr_convert() instead of crm_time_hr_new().

Reply:
Hi Ken,

Replacing already used crm_time_hr_new(), I think that it will be the final modification as below.
how is it?
(In order not to take different time, replace crm_time_hr_new().)

```
static int
exec_alert_list(lrmd_t *lrmd, GList *alert_list, enum crm_alert_flags kind,
                const char *attr_name, lrmd_key_value_t *params)
{   
    bool any_success = FALSE, any_failure = FALSE;
    const char *kind_s = crm_alert_flag2text(kind);
    crm_time_hr_t *now = NULL;
    struct timeval tv_now;
    char timestamp_epoch[34];
    char timestamp_usec[14];

(snip)
    for (GList *iter = g_list_first(alert_list); iter; iter = g_list_next(iter)) {
        crm_alert_entry_t *entry = (crm_alert_entry_t *)(iter->data);
        lrmd_key_value_t *copy_params = NULL;
        lrmd_key_value_t *head = NULL;
        int rc;

(snip)

        if (now == NULL) {
// ----> In order not to take different time, replace crm_time_hr_new ().
            if (gettimeofday(&tv_now, NULL) == 0){
                now = crm_time_timeval_hr_convert(NULL, &tv_now);
            }
        }
        crm_info("Sending %s alert via %s to %s",
                 kind_s, entry->id, entry->recipient);
(snip)
        if (now) {
            char *timestamp = crm_time_format_hr(entry->tstamp_format, now);

            if (timestamp) {
                copy_params = alert_key2param(copy_params, CRM_alert_timestamp,
                                              timestamp);
                free(timestamp);
            }

            snprintf(timestamp_epoch, 32, "%ld", tv_now.tv_sec);
            copy_params = alert_key2param(copy_params, CRM_alert_timestamp_epoch, 
                                      timestamp_epoch);
            snprintf(timestamp_usec, 12, "%06d", now->useconds);
            copy_params = alert_key2param(copy_params, CRM_alert_timestamp_msec, 
                                     timestamp_usec);
        }
(snip)
```

Reply:
Hi Hideo,

Yes, that looks good!

One minor change I would make is that timestamp_epoch only has to be [20] (maximum 19 decimal digits in a 64-bit signed integer plus a null terminator), and timestamp_usec only has to be [7]. The second argument to snprintf() (the maximum length) includes the null terminator, so you can use sizeof(timestamp_epoch) and sizeof(timestamp_usec).

Reply:
Hi Ken,

How about with the next fix?

```
(snip)
    char timestamp_epoch[20];
    char timestamp_usec[7];
(snip)
            snprintf(timestamp_epoch, sizeof(timestamp_epoch), "%ld", tv_now.tv_sec);
            copy_params = alert_key2param(copy_params, CRM_alert_timestamp_epoch, timestamp_epoch);
            snprintf(timestamp_usec, sizeof(timestamp_usec), "%06d", now->useconds);
            copy_params = alert_key2param(copy_params, CRM_alert_timestamp_ns, timestamp_usec);
(snip)

```

If there is no problem, I will make the last PR reflecting your pointed out.
There are a few circumstances, PR will be delayed.

Many thanks!
Hideo Yamauchi.

Reply:
Hi Hideo,

Yes, that looks good. There's no hurry -- I plan on releasing 2.0.0-rc2 today, but we still have plenty of time to get it into rc3.